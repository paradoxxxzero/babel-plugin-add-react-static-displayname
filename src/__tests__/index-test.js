// src/__tests__/index-test.js
const babel = require('@babel/core')
const plugin = require('..')

describe('sets displayName on simple class', () => {
  const clsSimple = `
class Simple {
  render() {
    return null
  }
}
`

  it('sets it', () => {
    const { code } = babel.transform(clsSimple, {
      plugins: ['@babel/plugin-syntax-class-properties', plugin],
    })
    expect(code).toMatchSnapshot()
  })

  it('is transformed with class-properties plugin', () => {
    const { code } = babel.transform(clsSimple, {
      plugins: [plugin, '@babel/plugin-proposal-class-properties'],
    })
    expect(code).toMatchSnapshot()
  })
})

describe('sets displayName on class with static members', () => {
  const clsWithStatic = `
class WithStatic {
  static attr = 2

  render() {
    return null
  }
}
`

  it('is added along with the other static', () => {
    const { code } = babel.transform(clsWithStatic, {
      plugins: ['@babel/plugin-syntax-class-properties', plugin],
    })
    expect(code).toMatchSnapshot()
  })

  it('is transformed with the other static', () => {
    const { code } = babel.transform(clsWithStatic, {
      plugins: [plugin, '@babel/plugin-proposal-class-properties'],
    })
    expect(code).toMatchSnapshot()
  })
})

describe('sets displayName on class with decorators', () => {
  const clsWithDecorator = `
@decorator
class WithDecorator {
  static attr = 2

  render() {
    return null
  }
}
`

  it('works with decorators', () => {
    const { code } = babel.transform(clsWithDecorator, {
      plugins: [
        '@babel/plugin-syntax-class-properties',
        plugin,
        '@babel/plugin-proposal-decorators',
      ],
    })
    expect(code).toMatchSnapshot()
  })

  it('works with decorators and class-properties', () => {
    const { code } = babel.transform(clsWithDecorator, {
      plugins: [
        '@babel/plugin-proposal-decorators',
        plugin,
        ['@babel/plugin-proposal-class-properties', { loose: true }],
      ],
    })
    expect(code).toMatchSnapshot()
  })
})

describe("does not set displayName if it's already set", () => {
  const clsWithAlreadyDisplayName = `
class WithAlreadyDisplayName {
  static displayName = 'Component'

  render() {
    return null
  }
}
`

  it('does not overide displayName', () => {
    const { code } = babel.transform(clsWithAlreadyDisplayName, {
      plugins: ['@babel/plugin-syntax-class-properties', plugin],
    })
    expect(code).toMatchSnapshot()
  })
})

describe("does not set displayName if there's no render method", () => {
  const clsNoRender = `
class NoRender {
  render_it() {
    return null
  }
}
`

  it('does not overide displayName', () => {
    const { code } = babel.transform(clsNoRender, {
      plugins: ['@babel/plugin-syntax-class-properties', plugin],
    })
    expect(code).toMatchSnapshot()
  })
})
