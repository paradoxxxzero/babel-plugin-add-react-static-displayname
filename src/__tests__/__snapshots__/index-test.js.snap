// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not set displayName if it's already set does not overide displayName 1`] = `
"class WithAlreadyDisplayName {
  static displayName = 'Component';

  render() {
    return null;
  }

}"
`;

exports[`does not set displayName if there's no render method does not overide displayName 1`] = `
"class NoRender {
  render_it() {
    return null;
  }

}"
`;

exports[`sets displayName on class with decorators works with decorators 1`] = `
"var _class;

let WithDecorator = decorator(_class = class WithDecorator {
  static displayName = \\"WithDecorator\\";
  static attr = 2;

  render() {
    return null;
  }

}) || _class;"
`;

exports[`sets displayName on class with decorators works with decorators and class-properties 1`] = `
"var _class, _class2, _temp;

let WithDecorator = decorator(_class = (_temp = _class2 = class WithDecorator {
  render() {
    return null;
  }

}, _class2.displayName = \\"WithDecorator\\", _class2.attr = 2, _temp)) || _class;"
`;

exports[`sets displayName on class with static members is added along with the other static 1`] = `
"class WithStatic {
  static displayName = \\"WithStatic\\";
  static attr = 2;

  render() {
    return null;
  }

}"
`;

exports[`sets displayName on class with static members is transformed with the other static 1`] = `
"class WithStatic {
  render() {
    return null;
  }

}

Object.defineProperty(WithStatic, \\"displayName\\", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: \\"WithStatic\\"
});
Object.defineProperty(WithStatic, \\"attr\\", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: 2
});"
`;

exports[`sets displayName on simple class is transformed with class-properties plugin 1`] = `
"class Simple {
  render() {
    return null;
  }

}

Object.defineProperty(Simple, \\"displayName\\", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: \\"Simple\\"
});"
`;

exports[`sets displayName on simple class sets it 1`] = `
"class Simple {
  static displayName = \\"Simple\\";

  render() {
    return null;
  }

}"
`;
